pipeline/mutate/mutator_hydrator.go

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright © 2017-2018 Aeneas Rekkas <aeneas+oss@aeneas.io>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @author       Aeneas Rekkas <aeneas+oss@aeneas.io>
+ * @copyright  2017-2018 Aeneas Rekkas <aeneas+oss@aeneas.io>
+ * @license  	   Apache-2.0
+ */
+
+package mutate
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/cenkalti/backoff"
+
+	"github.com/ory/x/httpx"
+
+	"github.com/pkg/errors"
+
+	"github.com/ory/oathkeeper/driver/configuration"
+	"github.com/ory/oathkeeper/pipeline"
+	"github.com/ory/oathkeeper/pipeline/authn"
+)
+
+const (
+	ErrMalformedResponseFromUpstreamAPI = "The call to an external API returned an invalid JSON object"
+	ErrMissingAPIURL                    = "Missing URL in mutator configuration"
+	ErrInvalidAPIURL                    = "Invalid URL in mutator configuration"
+	ErrNon200ResponseFromAPI            = "The call to an external API returned a non-200 HTTP response"
+	ErrInvalidCredentials               = "Invalid credentials were provided in mutator configuration"
+	ErrNoCredentialsProvided            = "No credentials were provided in mutator configuration"
+	defaultNumberOfRetries              = 3
+	defaultDelayInMilliseconds          = 100
+	contentTypeHeaderKey                = "Content-Type"
+	contentTypeJSONHeaderValue          = "application/json"
+)
+
+type MutatorHydrator struct {
+	c      configuration.Provider
+	client *http.Client
+}
+
+type BasicAuthn struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+type Authentication struct {
+	Basic BasicAuthn `json:"basic"`
+}
+
+type RetryConfig struct {
+	NumberOfRetries     int `json:"number"`
+	DelayInMilliseconds int `json:"delayInMilliseconds"`
+}
+
+type externalAPIConfig struct {
+	Url   string          `json:"url"`
+	Authn *Authentication `json:"authn,omitempty"`
```

Comment:
This would probably be `authorization` right? Because we're modifying HTTP Authorization?

Reply:
Sure, we are modifying `HTTP Authorization` header, but it is a header used for authentication. And the config is actually specifying [*Authentication* scheme](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#Authentication_schemes) and credentials. But if this argumentation doesn't convince you and you prefer it to be `Authorization` I will not argue to change that 😄

Reply:
We could also just change it to `auth`, then it implies both 😅

Reply:
Ok, changed it. Let me know what you think about it now 😄 
I will adjust documentation in a second